Sources/Basics/Concurrency/ConcurrencyHelpers.swift

```diff
@@ -76,3 +77,183 @@ extension DispatchQueue {
         }
     }
 }
+
+/// A queue for running async operations with a limit on the number of concurrent tasks.
+public final class AsyncOperationQueue: @unchecked Sendable {
+
+    // This implementation is identical to the AsyncOperationQueue in swift-build.
+    // Any modifications made here should also be made there.
+    // https://github.com/swiftlang/swift-build/blob/main/Sources/SWBUtil/AsyncOperationQueue.swift#L13
+
+    fileprivate typealias ID = UUID
+    fileprivate typealias WaitingContinuation = CheckedContinuation<Void, any Error>
+
+    private let concurrentTasks: Int
+    private var activeTasks: Int = 0
+    private var waitingTasks: [WaitingTask] = []
+    private let waitingTasksLock = NSLock()
+
+    fileprivate enum WaitingTask {
+        case creating(ID)
+        case waiting(ID, WaitingContinuation)
+        case cancelled(ID)
+
+        var id: ID {
+            switch self {
+            case .creating(let id), .waiting(let id, _), .cancelled(let id):
+                return id
+            }
+        }
+
+        var continuation: WaitingContinuation? {
+            guard case .waiting(_, let continuation) = self else {
+                return nil
+            }
+            return continuation
+        }
+    }
+
+    /// Creates an `AsyncOperationQueue` with a specified number of concurrent tasks.
+    /// - Parameter concurrentTasks: The maximum number of concurrent tasks that can be executed concurrently.
+    public init(concurrentTasks: Int) {
+        self.concurrentTasks = concurrentTasks
+    }
+
+    deinit {
+        waitingTasksLock.withLock {
+            if !waitingTasks.filter({ $0.continuation != nil }).isEmpty {
```

Comment:
Isn't it unexpected if there is _anything at all_ in waitingTasks, not just those with a non-nil continuation?